design/external-authorization-design.md

```diff
@@ -0,0 +1,453 @@
+# External Authorization Support
+
+Status: Reviewing
+
+## Abstract
+This document describes a design for performing request authorization for virtual hosts hosted by Contour.
+
+## Background
+
+- [#432][5] Provide support for integrating with an external auth service.
+- [#2459][3] External auth background and goals.
+
+## Goals
+- Allow operators to integrate existing authorization services with Contour.
+- Support arbitrary types of authentication and authorization.
+- Decouple Contour from external authorization, so it can evolve at an independent rate.
+- Integrate cleanly with the HTTPProxy API.
+
+## Non Goals
+
+- Support for multiple external authorization mechanisms.
+  Various proxies have different mechanisms for integrating external services, but it is impractical to support them all.
+
+- Abstracting Envoy authorization protocols.
+  The scope of abstracting the Envoy authorization protocol is too large to be tractable.
+  Abstracting the protocol would also work against the goal of being able to integrate existing authorization servers.
+
+## High-Level Design
+A new CRD `SupportService` adds a way to represent and track an authorization service.
```

Comment:
I think we should just build out a proper CRD for Auth and not try to create a generic one that is useful for many purposes. I feel in the long run, we're going to get into a spot where it's super difficult to setup since it's so generic.

Reply:
I don't think it's tractable to build that. There are an unbounded set of different authorization mechanisms and trying to get them all into a single CRD will be a mess. It has the other disadvantage that it's not extensible by operators, so that only Contour maintainers can implement new authorization types.